server/src/main/java/org/opensearch/index/ShardIndexingPressureStore.java

```diff
@@ -0,0 +1,121 @@
+        this.maxColdStoreSize = MAX_COLD_STORE_SIZE.get(settings).intValue();
+        clusterSettings.addSettingsUpdateConsumer(MAX_COLD_STORE_SIZE, this::setMaxColdStoreSize);
+    }
+
+    public ShardIndexingPressureTracker getShardIndexingPressureTracker(ShardId shardId) {
+        ShardIndexingPressureTracker tracker = shardIndexingPressureHotStore.get(shardId);
+        if (isNull(tracker)) {
+            // Attempt from Indexing pressure cold store
+            tracker = shardIndexingPressureColdStore.get(shardId);
+            // If not already present in cold store instantiate a new one
+            if (isNull(tracker)) {
+                ShardIndexingPressureTracker newShardIndexingPressureTracker = new ShardIndexingPressureTracker(shardId,
+                    this.shardIndexingPressureSettings.getShardPrimaryAndCoordinatingBaseLimits(),
+                    this.shardIndexingPressureSettings.getShardReplicaBaseLimits());
+                // Try update the new shard tracker to the hot store
+                tracker = shardIndexingPressureHotStore.putIfAbsent(shardId, newShardIndexingPressureTracker);
+                // Update the tracker so that we use the one actual in the hot store
+                tracker = tracker == null ? newShardIndexingPressureTracker : tracker;
+                // Write through into the cold store for future reference
+                updateShardIndexingPressureColdStore(tracker);
+            } else {
+                // Attempt update tracker to the hot store and return the tracker which finally made to the hot store to avoid any race
+                ShardIndexingPressureTracker newTracker = shardIndexingPressureHotStore.putIfAbsent(shardId, tracker);
+                tracker = newTracker == null ? tracker : newTracker;
+            }
+        }
+        return tracker;
+    }
+
+    public Map<ShardId, ShardIndexingPressureTracker> getShardIndexingPressureHotStore() {
+        return Collections.unmodifiableMap(shardIndexingPressureHotStore);
+    }
+
+    public Map<ShardId, ShardIndexingPressureTracker> getShardIndexingPressureColdStore() {
+        return Collections.unmodifiableMap(shardIndexingPressureColdStore);
+    }
+
+    public void tryTrackerCleanupFromHotStore(ShardIndexingPressureTracker tracker, Supplier<Boolean> condition) {
+        if (condition.get()) {
+            // Try inserting into cold store again in case there was an eviction triggered
+            shardIndexingPressureColdStore.putIfAbsent(tracker.getShardId(), tracker);
+            // Remove from the hot store
+            shardIndexingPressureHotStore.remove(tracker.getShardId(), tracker);
+        }
+    }
+
+    private void updateShardIndexingPressureColdStore(ShardIndexingPressureTracker tracker) {
+        if (shardIndexingPressureColdStore.size() > maxColdStoreSize) {
+            shardIndexingPressureColdStore.clear();
+        }
+        shardIndexingPressureColdStore.put(tracker.getShardId(), tracker);
+    }
```

Comment:
There seems to be a concurrency bug here, you'll end up clearing the store multiple times and losing the previous entry after clear. Hence please follow the suggestion as made above

Reply:
1. It is not a bug but an intended approach to intentionally clear up trackers, once they are no more in use, and allows us to clear up the memory. In most indexing scenarios, indexes are expected to rotate, and hence the new trackers will be created per shard to track the new operations. Tracker if not present in a ``host store``, will be eventually removed from the cold store once the purge happens. Trackers in hot store will be updated back to cold upon request completion. 
So Cold essentially acts a transient store between successive operations for the same shard which are at some gap. At any point customers would use the hot trackers to track the current cluster performance, stuck tasks, host shards etc. Size of cold store is of significance to avoid unnecessary churns, and is provided as a dynamic setting for tuning.
 
2. It is important to size the ``shard_indexing_pressure.cache_store.max_size`` setting correctly, to ensure unnecessary purge doesn't happen for the number of indexes which are being concurrently updated. LRU could be an alternative to full flush, but is more expensive. So choosing simplicity currently as it serves the purpose well. Added a note to revisit this later, evaluate, and change to LRU if need be.

Reply:
Don't think you got my point here, Lets discuss more

Reply:
I see your concerns around race between the threads trying to clear up the state simultaneously. However I see that as a tradeoff, to keep the approach less synchornized and simplistic. Flush is a less frequent operation, can be sized based on workload, and it is okay to not have strong consistency on counters (under such rare races), as long as we are able to address majority of use cases at ease. 
Yes, let's take walk a few scenarios and discuss this more.

Reply:
What is confusing is you use cold store as a collection with aggressive concurrency when you don't need this based on why you have mentioned.

Reply:
It is required for a scenario when large number of gets are requested concurrently for a tracker which either (1) do not exist and needs to be created and updated (2) exists in cold store and needs to be added to hot store. From functional use case its traffic spike for a relatively new shard or a completely new shard (index rollover).

Reply:
Please add a clarifying comment that such discrepancies are fine

Reply:
done